### PR TITLE
Flink: Enable multiple flink sinks for the same table in the same job

### PR DESCRIPTION
**Pull request by pvary**
Currently the Flink Iceberg commits are identifier by the `table` and the `JobID`. This could become problematic when there are multiple sinks for the same table for the same job.

Let's imagine the following sequence of events:
- Sink1 `notifyCheckpointComplete(99)` commits checkpoint the changes to the Iceberg table, and marks `99` as the latest checkpoint which was committed
- Sink1 `notifyCheckpointComplete(99)` commits checkpoint the changes to the Iceberg table, and marks `99` as the latest checkpoint which was committed
- New checkpoint is created with the id `100`
- Sink1 `notifyCheckpointComplete(100)` commits checkpoint the changes to the Iceberg table, and marks `100` as the latest checkpoint which was committed
- Job is cancelled
- Job is restarted
- Sink1 sees that the latest checkpoint which is committed is `100`, so nothing to do
- Sink2 sees that the latest checkpoint which is committed is `100`, so nothing to do - Which is wrong as Sink2 never committed the relevant changes to the table

To fix this added another identifier `operatorUniqueId` is added to the snapshot summary when checking for the commits, and this operatorId is also checked on recovery.

The PR contains 2 additional unit tests to check the scenario.

### CODE REVIEW

File: flink/v1.16/flink/src/main/java/org/apache/iceberg/flink/sink/IcebergFilesCommitter.java [hunk 1]
```diff
@@ -134,6 +136,7 @@ class IcebergFilesCommitter extends AbstractStreamOperator<Void>
   public void initializeState(StateInitializationContext context) throws Exception {
     super.initializeState(context);
     this.flinkJobId = getContainingTask().getEnvironment().getJobID().toString();
+    this.operatorUniqueId = getRuntimeContext().getOperatorUniqueID();
```
Comment (stevenzwu) [inline flink/v1.16/flink/src/main/java/org/apache/iceberg/flink/sink/IcebergFilesCommitter.java:139]: This can be a problem. it gets the `jobVertexId` (a UUID like string). I verified it by setting the `uidPrefix=iceberg-sink` in `TestFlinkIcebergSink`.
```
    FlinkSink.forRow(dataStream, SimpleDataUtil.FLINK_SCHEMA)
        ...
        .uidPrefix("iceberg-sink")
        .append();
````

Then I can see the `operatorUniqueId` value as jobVertexId of `0e4eaa3db14ae3b6752b25172fd05c50`. It can change when the job graph changed (e.g. adding a stateless map operator).  Then we won't be able to match the operator id properly. 

Ideally, we want to use the [operator uid from Flink](https://nightlies.apache.org/flink/flink-docs-release-1.16/docs/ops/upgrading/#datastream-api), which can be set by users. See the code from `FlinkSink`. 
```
      if (uidPrefix != null) {
        committerStream = committerStream.uid(uidPrefix + "-committer");
      }
```

if uid is not provided by application, Flink automatically generates a UUID by default. Setting uid for stateful operator (like Iceberg committer) is a [best practice](https://nightlies.apache.org/flink/flink-docs-release-1.16/docs/ops/production_ready/#set-uuids-for-all-operators). Flink uses uid to match state description. if uid changed, old state won't be matched and restored. uid works perfectly for this purpose. But I am not sure what is the best way to retrieve the `uid` in the `initializeState` method. 

I can see the operator name from debugger, which is also set by `FlinkSink` based on `uidPrefix`. It will be unique as `uidPrefix` should be unique. Nonetheless, it is still not the ideal candidate/replacement for `uid` though.
image

File: flink/v1.16/flink/src/test/java/org/apache/iceberg/flink/sink/TestIcebergFilesCommitter.java [hunk 2]
```diff
@@ -554,21 +579,178 @@ public void testMultipleJobsWriteSameTable() throws Exception {
...
+      SimpleDataUtil.assertTableRows(table, ImmutableList.of(row1));
+      assertSnapshotSize(1);
+      assertMaxCommittedCheckpointId(jobId, operatorId1, checkpointId);
+      assertMaxCommittedCheckpointId(jobId, operatorId2, -1);
+    }
+
+    // Restore from the given snapshot
+    try (OneInputStreamOperatorTestHarness<WriteResult, Void> harness1 = createStreamSink(jobId);
+        OneInputStreamOperatorTestHarness<WriteResult, Void> harness2 = createStreamSink(jobId)) {
+      harness1.getStreamConfig().setOperatorID(operatorId1);
```
Comment (stevenzwu) [inline flink/v1.16/flink/src/test/java/org/apache/iceberg/flink/sink/TestIcebergFilesCommitter.java:697]: this didn't test the case where operatorId/jobVertexId can change.

File: flink/v1.16/flink/src/main/java/org/apache/iceberg/flink/sink/IcebergFilesCommitter.java (hunk 1 above)
Comment (stevenzwu) [reply flink/v1.16/flink/src/main/java/org/apache/iceberg/flink/sink/IcebergFilesCommitter.java:139]: I did more experiment. if the `uidPrefix` is set in `FlinkSink`, then it seems that `getOperatorUniqueID`  does return a stable ID even if the job DAG changed.

if `uidPrefix` is not set in `FlinkSink`, DAG change leads to unstable/different values from `getOperatorUniqueID`. This is not really a problem, because we should always set `uidPrefix` for `FlinkSink` as best practice for stateful committer operator.

In summary, using `getOperatorUniqueID` seems acceptable to me.

Comment (stevenzwu) [reply flink/v1.16/flink/src/main/java/org/apache/iceberg/flink/sink/IcebergFilesCommitter.java:139]: cc @hililiwei @openinx and see if they have other suggestions.

File: flink/v1.16/flink/src/test/java/org/apache/iceberg/flink/sink/TestIcebergFilesCommitter.java [hunk 3]
```diff
@@ -554,21 +579,178 @@ public void testMultipleJobsWriteSameTable() throws Exception {
...
+  public void testMultipleSinksWriteSameTable() throws Exception {
+    long timestamp = 0;
+    List<RowData> tableRows = Lists.newArrayList();
+
+    JobID[] jobs = new JobID[] {new JobID(), new JobID(), new JobID()};
+    OperatorID[] operatorIds1 =
+        new OperatorID[] {new OperatorID(), new OperatorID(), new OperatorID()};
+    OperatorID[] operatorIds2 =
+        new OperatorID[] {new OperatorID(), new OperatorID(), new OperatorID()};
+    for (int i = 0; i < 20; i++) {
```
Comment (stevenzwu) [inline flink/v1.16/flink/src/test/java/org/apache/iceberg/flink/sink/TestIcebergFilesCommitter.java:597]: why do we need a for loop over multiple jobs here? a single job is probably good for this purpose?

File: flink/v1.16/flink/src/test/java/org/apache/iceberg/flink/sink/TestIcebergFilesCommitter.java [hunk 4]
```diff
@@ -554,21 +579,178 @@ public void testMultipleJobsWriteSameTable() throws Exception {
...
+
+      assertSnapshotSize(0);
+      assertMaxCommittedCheckpointId(jobId, operatorId1, -1L);
+      assertMaxCommittedCheckpointId(jobId, operatorId2, -1L);
+
+      RowData row1 = SimpleDataUtil.createRowData(1, "hello1");
+      expectedRows.add(row1);
+      DataFile dataFile1 = writeDataFile("data-1-1", ImmutableList.of(row1));
+
+      harness1.processElement(of(dataFile1), ++timestamp);
```
Comment (stevenzwu) [inline flink/v1.16/flink/src/test/java/org/apache/iceberg/flink/sink/TestIcebergFilesCommitter.java:672]: I remember feedbacks on avoid using `++` with side effect. Iceberg code style prefers separate increment action after value read.

File: flink/v1.16/flink/src/test/java/org/apache/iceberg/flink/sink/TestIcebergFilesCommitter.java (hunk 3 above)
Comment (stevenzwu) [reply flink/v1.16/flink/src/test/java/org/apache/iceberg/flink/sink/TestIcebergFilesCommitter.java:597]: this test seems redundant given the coverage from the test method below

File: flink/v1.16/flink/src/test/java/org/apache/iceberg/flink/sink/TestIcebergFilesCommitter.java [hunk 5]
```diff
@@ -554,21 +579,178 @@ public void testMultipleJobsWriteSameTable() throws Exception {
...
+
+    // Restore from the given snapshot
+    try (OneInputStreamOperatorTestHarness<WriteResult, Void> harness1 = createStreamSink(jobId);
+        OneInputStreamOperatorTestHarness<WriteResult, Void> harness2 = createStreamSink(jobId)) {
+      harness1.getStreamConfig().setOperatorID(operatorId1);
+      harness1.setup();
+      harness1.initializeState(snapshot1);
+      harness1.open();
+
+      harness2.getStreamConfig().setOperatorID(operatorId2);
```
Comment (stevenzwu) [inline flink/v1.16/flink/src/test/java/org/apache/iceberg/flink/sink/TestIcebergFilesCommitter.java:702]: maybe we should have another test for the case where operatorId changed.

File: flink/v1.16/flink/src/main/java/org/apache/iceberg/flink/sink/IcebergFilesCommitter.java (hunk 1 above)
Comment (gyfora) [reply flink/v1.16/flink/src/main/java/org/apache/iceberg/flink/sink/IcebergFilesCommitter.java:139]: `operatorUniqueId` is deterministaclly generated based on the uid of the operator which is used already by Flink to assign the state to this operator. This is competely correct.

**Review by gyfora — APPROVED** [review]
lgtm +1

File: flink/v1.16/flink/src/test/java/org/apache/iceberg/flink/sink/TestIcebergFilesCommitter.java (hunk 3 above)
Comment (pvary) [reply flink/v1.16/flink/src/test/java/org/apache/iceberg/flink/sink/TestIcebergFilesCommitter.java]: I was thinking about adding a test case with the happy path, but you are right that we already cover that in the test below.

Removed

File: flink/v1.16/flink/src/test/java/org/apache/iceberg/flink/sink/TestIcebergFilesCommitter.java (hunk 4 above)
Comment (pvary) [reply flink/v1.16/flink/src/test/java/org/apache/iceberg/flink/sink/TestIcebergFilesCommitter.java:615]: I think changing
```
      harness1.processElement(of(dataFile1), ++timestamp);
```
to
```
      harness1.processElement(of(dataFile1), timestamp);
      timestamp++;
```
would not improve the readability of the code. It would make it even worse.
Also the whole class is full of these already (43 occurrences without the new ones). Removing them only here would be awkward.

File: flink/v1.16/flink/src/test/java/org/apache/iceberg/flink/sink/TestIcebergFilesCommitter.java (hunk 5 above)
Comment (pvary) [reply flink/v1.16/flink/src/test/java/org/apache/iceberg/flink/sink/TestIcebergFilesCommitter.java:645]: I might be misunderstand something.
We need the operatorId to be constant, and static. If it is changing, then we will have issues with the recovery as all of the previously collected, but not committed changes will recommitted.

Shall we make the setting of the uuid mandatory, and throw an exception if it is not set on a FlinkSink?

Comment (stevenzwu) [reply flink/v1.16/flink/src/test/java/org/apache/iceberg/flink/sink/TestIcebergFilesCommitter.java:645]: > If it is changing, then we will have issues with the recovery as all of the previously collected, but not committed changes will be lost (as the state is lost for the operator). 

I was thinking about a test to demonstrate this effect.

File: flink/v1.16/flink/src/test/java/org/apache/iceberg/flink/sink/TestIcebergFilesCommitter.java (hunk 4 above)
Comment (stevenzwu) [reply flink/v1.16/flink/src/test/java/org/apache/iceberg/flink/sink/TestIcebergFilesCommitter.java:615]: fair enough if it is already used in this class.

this was a feedback I got from Ryan before.

File: flink/v1.16/flink/src/main/java/org/apache/iceberg/flink/sink/IcebergFilesCommitter.java [hunk 6]
```diff
@@ -69,6 +69,7 @@ class IcebergFilesCommitter extends AbstractStreamOperator<Void>
 
   private static final Logger LOG = LoggerFactory.getLogger(IcebergFilesCommitter.class);
   private static final String FLINK_JOB_ID = "flink.job-id";
+  private static final String OPERATOR_UNIQUE_ID = "flink.operator-unique-id";
```
Comment (stevenzwu) [inline flink/v1.16/flink/src/main/java/org/apache/iceberg/flink/sink/IcebergFilesCommitter.java:72]: nit: would `flink.operator-id` be sufficient? it also matches the `job-id` style. `id` usually implies unique.

Comment (stevenzwu) [reply flink/v1.16/flink/src/main/java/org/apache/iceberg/flink/sink/IcebergFilesCommitter.java:72]: similarly, the constant name can be `FLINK_OPERATOR_ID`

File: flink/v1.16/flink/src/test/java/org/apache/iceberg/flink/sink/TestIcebergFilesCommitter.java (hunk 5 above)
Comment (stevenzwu) [reply flink/v1.16/flink/src/test/java/org/apache/iceberg/flink/sink/TestIcebergFilesCommitter.java:645]: never mind. if operator id changed, the state won't be restored anyway and hence there is nothing to restore.

**Review by stevenzwu — APPROVED** [review]
LGTM. left a nit comment on the naming for consistency.

**Comment by pvary** [comment]
Merged the PR.
Thanks for the review @stevenzwu  and @gyfora!

**Comment by chenwyi2** [comment]
there is a problem, since the key of checkpoint state is unique, but two sinks will have the same checkpoint id？

**Comment by pvary** [comment]
@chenwyi2: 2 sinks have a different operatorid
